src/upload.rs

```diff
@@ -0,0 +1,215 @@
+use std::path::PathBuf;
+use std::process::Command;
+use tar::Archive;
+use thiserror::Error;
+use uuid::Uuid;
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum UploadError {
+    #[error("The project is too large to be uploaded.")]
+    TooLarge,
+    #[error("Failed to save zip file.")]
+    SaveFile,
+    #[error("Failed to copy files.")]
+    CopyFiles,
+    #[error("Invalid Forc version: {0}")]
+    InvalidForcVersion(String),
+    #[error("Failed to compile project.")]
+    FailedToCompile,
+    #[error("Failed to authenticate.")]
+    Authentication,
+    #[error("Failed to upload to IPFS.")]
+    Ipfs,
+    #[error("Failed to generate bytecode ID. Err: {0}")]
+    BytecodeId(String),
+}
+
+pub async fn handle_project_upload(
+    upload_dir: &Path,
+    upload_id: &Uuid,
+    orig_tarball_path: &PathBuf,
+    forc_path: &Path,
+    forc_version: String,
+    pinata_client: &impl PinataClient,
+) -> Result<NewUpload, UploadError> {
+    let unpacked_dir = upload_dir.join("unpacked");
+    let release_dir = unpacked_dir.join("out/release");
+    let project_dir = upload_dir.join("project");
+
+    // Unpack the tarball.
+    let tarball = File::open(orig_tarball_path).unwrap();
+    let decompressed = GzDecoder::new(tarball);
+    let mut archive = Archive::new(decompressed);
+    archive.unpack(&unpacked_dir).unwrap();
+
+    // Remove `out` directory if it exists.
+    let _ = fs::remove_dir_all(unpacked_dir.join("out"));
+
+    let output = Command::new(format!("{}/bin/forc", forc_path.to_string_lossy()))
+        .arg("build")
+        .arg("--release")
+        .current_dir(&unpacked_dir)
+        .output()
+        .expect("Failed to execute forc build");
+
+    if !output.status.success() {
+        return Err(UploadError::FailedToCompile);
+    }
+
+    // Copy files that are part of the Sway project to a new directory.
+    let output = Command::new("rsync")
+        .args([
+            "-av",
+            "--prune-empty-dirs",
+            "--include=*/",
+            "--include=Forc.toml",
+            "--include=Forc.lock",
+            "--include=*.sw",
+            "--exclude=*",
+            "unpacked/",
+            "project",
+        ])
+        .current_dir(upload_dir)
+        .output()
+        .expect("Failed to copy project files");
+
+    if !output.status.success() {
+        return Err(UploadError::CopyFiles);
+    }
+
+    // Pack the new tarball.
+    let final_tarball_path = upload_dir.join("project.tgz");
+    let tar_gz = File::create(&final_tarball_path).unwrap();
+    let enc = GzEncoder::new(tar_gz, Compression::default());
+    let mut tar = tar::Builder::new(enc);
+
+    // Add files to the tar archive
+    tar.append_dir_all(".", &project_dir).unwrap();
+
+    // Finish writing the tar archive
+    tar.finish().unwrap();
+
+    // Make sure the GzEncoder finishes and flushes all data
+    let enc = tar.into_inner().unwrap();
+    enc.finish().unwrap();
+
+    // Store the tarball in IPFS.
```

Comment:
Can we also upload these to S3 in addition to IPFS?

Protocol Labs has been having issues keeping their IPFS indexers online, and based on performance issues we had at Infura, I don't feel comfortable trusting our only copy of this data to a protocol that doesn't have the best uptime record.

Reply:
Possibly [R2 from cloudflare](https://www.cloudflare.com/en-gb/developer-platform/products/r2/) could be a good alternative.

Reply:
I haven't looked, but there's probably a library with a standard interface that does all of these backends and more.

Reply:
Great feedback, I'll do this in a separate PR: https://github.com/FuelLabs/forc.pub/issues/20